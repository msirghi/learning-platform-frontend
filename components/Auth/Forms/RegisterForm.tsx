import React, { useState } from 'react';
import { Button, LinearProgress, Typography } from '@material-ui/core';
import styles from '../../../styles/modules/Auth.module.scss';
import { useTranslation } from '../../../i18n';
import ValidationService from '../../../services/ValidationService';
import { useForm } from 'react-hook-form';
import TextField from '@material-ui/core/TextField';
import RegexContants from '../../../common/constants/regex.constants';
import Checkbox from '@material-ui/core/Checkbox';
import FormControlLabel from '@material-ui/core/FormControlLabel';

interface Values {
  email: string;
  password: string;
  repeatPassword: string;
  firstName: string;
  lastName: string;
  checked: boolean | string;
}

function RegisterForm() {
  const { t } = useTranslation();
  const [isSubmitting, setSubmitting] = useState<boolean>(false);
  const { register, handleSubmit, errors, getValues } = useForm<Values>();

  const onSubmit = (values: Values) => {
    setSubmitting(true);
    setTimeout(() => setSubmitting(false), 500);
  };

  return (
    <form noValidate onSubmit={handleSubmit(onSubmit)}>
      {errors.checked && (
        <Typography color='secondary'>{t('auth:errorMessages.acceptTerms')}</Typography>
      )}

      {isSubmitting && (
        <div data-testid='progress'>
          <LinearProgress />
        </div>
      )}

      <TextField
        variant='outlined'
        name='email'
        label={t('auth:email')}
        inputProps={{
          ref: register({
            required: String(t('auth:errorMessages.required')),
            pattern: {
              value: RegexContants.EMAIL_REGEX,
              message: t('auth:errorMessages.invalidEmail')
            }
          }),
          'data-testid': 'email-field',
          name: 'email'
        }}
        data-test='email'
        fullWidth
        margin='normal'
        error={!!errors.email}
        helperText={errors.email && errors.email.message}
      />

      <TextField
        variant='outlined'
        type='text'
        label={t('auth:firstName')}
        name='firstName'
        error={!!errors.firstName}
        helperText={errors.firstName && errors.firstName.message}
        inputProps={{
          ref: register({
            required: String(t('auth:errorMessages.required')),
            pattern: {
              value: RegexContants.NAME_REGEX,
              message: t('auth:errorMessages.invalidName')
            }
          }),
          'data-testid': 'firstName-field'
        }}
        fullWidth
        margin='normal'
      />

      <TextField
        variant='outlined'
        type='text'
        label={t('auth:lastName')}
        name='lastName'
        fullWidth
        margin='normal'
        error={!!errors.lastName}
        helperText={errors.lastName && errors.lastName.message}
        inputProps={{
          ref: register({
            required: String(t('auth:errorMessages.required')),
            pattern: {
              value: RegexContants.NAME_REGEX,
              message: t('auth:errorMessages.invalidName')
            }
          }),
          'data-testid': 'lastName-field'
        }}
      />
      <TextField
        error={!!errors.password}
        helperText={
          (errors.password &&
            errors.password.type === 'validate' &&
            t('auth:errorMessages.weakPassword')) ||
          (errors.password && errors.password.message)
        }
        variant='outlined'
        type='password'
        label={t('auth:password')}
        name='password'
        fullWidth
        margin='normal'
        inputProps={{
          ref: register({
            required: String(t('auth:errorMessages.required')),
            validate: (val) => !ValidationService.isPasswordWeak(val)
          }),
          'data-testid': 'password-field'
        }}
      />

      <TextField
        error={!!errors.repeatPassword}
        helperText={
          (errors.repeatPassword &&
            errors.repeatPassword.type === 'validate' &&
            t('auth:errorMessages.passwordMismatch')) ||
          (errors.repeatPassword && errors.repeatPassword.message)
        }
        variant='outlined'
        type='password'
        label={t('auth:repeatPassword')}
        name='repeatPassword'
        fullWidth
        margin='normal'
        inputProps={{
          ref: register({
            required: String(t('auth:errorMessages.required')),
            validate: (val) => getValues().password === val
          }),
          'data-testid': 'repeatPassword-field'
        }}
      />

      <FormControlLabel
        control={
          <Checkbox
            inputProps={{ 'data-testid': 'checkbox' }}
            inputRef={register({ required: true })}
            name='checked'
            color={'primary'}
          />
        }
        label={String(t('auth:acceptTermsAndCondition'))}
      />

      <div className={styles.submitButton}>
        <Button
          type={'submit'}
          variant='contained'
          color='primary'
          disabled={isSubmitting}
          data-testid='submit-button'
          data-test='submit-button'
          className={styles.registerButton}
        >
          {t('auth:registerButtonLabel')}
        </Button>
      </div>
    </form>
  );
}

RegisterForm.getInitialProps = async () => {
  return {
    namespacesRequired: ['auth']
  };
};

export default RegisterForm;
