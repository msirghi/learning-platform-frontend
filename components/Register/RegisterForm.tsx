import * as React from 'react';
import { Formik, Form, Field } from 'formik';
import { Button, LinearProgress } from '@material-ui/core';
import { TextField } from 'formik-material-ui';
import Grid from '@material-ui/core/Grid';
import styles from '../../styles/Auth.module.scss';
import { CheckboxWithLabel } from 'formik-material-ui';
import { withTranslation } from '../../i18n';

interface Values {
  email: string;
  password: string;
}

function RegisterForm({ t }) {
  return (
    <Formik
      initialValues={{
        email: '',
        password: '',
      }}
      validate={(values) => {
        const errors: Partial<Values> = {};
        if (!values.email) {
          errors.email = 'Required';
        } else if (
          !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)
        ) {
          errors.email = 'Invalid email address';
        }
        return errors;
      }}
      onSubmit={(values, { setSubmitting }) => {
        setTimeout(() => {
          setSubmitting(false);
          alert(JSON.stringify(values, null, 2));
        }, 500);
      }}
    >
      {({ submitForm, isSubmitting }) => (
        <Form>
          <Field
            variant='outlined'
            component={TextField}
            name='email'
            type='email'
            label={t('auth:email')}
            fullWidth
            margin='normal'
          />
          <Field
            variant='outlined'
            component={TextField}
            type='text'
            label={t('auth:firstName')}
            name='firstName'
            fullWidth
            margin='normal'
          />
          <Field
            variant='outlined'
            component={TextField}
            type='text'
            label={t('auth:lastName')}
            name='lastName'
            fullWidth
            margin='normal'
          />
          <Field
            variant='outlined'
            component={TextField}
            type='password'
            label={t('auth:password')}
            name='password'
            fullWidth
            margin='normal'
          />
          <Field
            variant='outlined'
            component={TextField}
            type='password'
            label={t('auth:repeatPassword')}
            name='repeatPassword'
            fullWidth
            margin='normal'
          />

          <Field
            component={CheckboxWithLabel}
            color={'primary'}
            type='checkbox'
            name='checked'
            Label={{ label: t('auth:acceptTermsAndCondition') }}
          />

          {isSubmitting && <LinearProgress />}
          <div className={styles.submitButton}>
            <Button
              variant='contained'
              color='primary'
              disabled={isSubmitting}
              onClick={submitForm}
            >
              {t('auth:registerButtonLabel')}
            </Button>
          </div>
        </Form>
      )}
    </Formik>
  );
}

RegisterForm.getInitialProps = async () => {
  return {
    namespacesRequired: ['auth'],
  };
};

export default withTranslation()(RegisterForm);
