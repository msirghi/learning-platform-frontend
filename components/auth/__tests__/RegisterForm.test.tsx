import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import RegisterForm from '../forms/RegisterForm';

describe('RegisterForm component', () => {
  const defaultProps = {
    onTabChange: jest.fn(),
    onMessage: jest.fn()
  };

  it('should match snapshot', () => {
    const component = shallow(<RegisterForm {...defaultProps} />);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render all the fields properly', () => {
    const { queryByTestId } = render(<RegisterForm {...defaultProps} />);

    const emailField = queryByTestId('email-field');
    expect(emailField).toBeInTheDocument();

    const firstNameField = queryByTestId('firstName-field');
    expect(firstNameField).toBeInTheDocument();

    const lastNameField = queryByTestId('lastName-field');
    expect(lastNameField).toBeInTheDocument();

    const passwordField = queryByTestId('password-field');
    expect(passwordField).toBeInTheDocument();

    const repeatPasswordField = queryByTestId('repeatPassword-field');
    expect(repeatPasswordField).toBeInTheDocument();

    const checkbox = queryByTestId('checkbox');
    expect(checkbox).toBeInTheDocument();

    const submitButton = queryByTestId('submit-button');
    expect(submitButton).toBeInTheDocument();
  });

  it('should have submit button enabled by default', () => {
    const { queryByTestId } = render(<RegisterForm {...defaultProps} />);

    const submitButton = queryByTestId('submit-button');
    expect(submitButton).toBeInTheDocument();
    expect(submitButton).toBeEnabled();
  });

  it('should have enabled submit button on valid form', () => {
    const { queryByTestId } = render(<RegisterForm {...defaultProps} />);

    const emailField = queryByTestId('email-field');
    const firstNameField = queryByTestId('firstName-field');
    const lastNameField = queryByTestId('lastName-field');
    const passwordField = queryByTestId('password-field');
    const repeatPasswordField = queryByTestId('repeatPassword-field');
    const checkbox = queryByTestId('checkbox');
    const submitButton = queryByTestId('submit-button');

    fireEvent.change(emailField, { target: { value: 'test@mail.com' } });
    fireEvent.change(firstNameField, { target: { value: 'test' } });
    fireEvent.change(lastNameField, { target: { value: 'test' } });
    fireEvent.change(passwordField, { target: { value: 'test123' } });
    fireEvent.change(repeatPasswordField, { target: { value: 'test123' } });
    fireEvent.click(checkbox);

    expect(emailField).toHaveValue('test@mail.com');
    expect(submitButton).toBeEnabled();
  });

  it('should submit the form if values are valid', async () => {
    const { queryByTestId, getByRole } = render(<RegisterForm {...defaultProps} />);

    const emailField = queryByTestId('email-field');
    const firstNameField = queryByTestId('firstName-field');
    const lastNameField = queryByTestId('lastName-field');
    const passwordField = queryByTestId('password-field');
    const repeatPasswordField = queryByTestId('repeatPassword-field');
    const checkbox = getByRole('checkbox');
    const submitButton = queryByTestId('submit-button');

    fireEvent.change(emailField, { target: { value: 'test@mail.com' } });
    fireEvent.change(firstNameField, { target: { value: 'test' } });
    fireEvent.change(lastNameField, { target: { value: 'test' } });
    fireEvent.change(passwordField, { target: { value: 'test123' } });
    fireEvent.change(repeatPasswordField, { target: { value: 'test123' } });
    fireEvent.click(checkbox);
    fireEvent.click(submitButton);

    expect(emailField).toHaveValue('test@mail.com');
    expect(submitButton).toBeEnabled();
    expect(checkbox).toBeChecked();
  });

  it('should display error message on unchecked checkbox', () => {
    const { queryByTestId, getByText } = render(<RegisterForm {...defaultProps} />);
    const submitButton = queryByTestId('submit-button');
    fireEvent.click(submitButton);

    expect(submitButton).toBeEnabled();
    expect(getByText('auth:acceptTermsAndCondition')).toBeDefined();
  });

  it('should get value from getInitialProps', async () => {
    const props = await RegisterForm.getInitialProps();
    expect(props).toStrictEqual({
      namespacesRequired: ['auth']
    });
  });
});
