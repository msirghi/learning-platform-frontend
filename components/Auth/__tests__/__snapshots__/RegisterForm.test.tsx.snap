// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm component should match snapshot 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <WithStyles(ForwardRef(TextField))
    data-test="email"
    error={false}
    fullWidth={true}
    inputProps={
      Object {
        "data-testid": "email-field",
        "name": "email",
        "ref": [Function],
      }
    }
    label="auth:email"
    margin="normal"
    name="email"
    variant="outlined"
  />
  <WithStyles(ForwardRef(TextField))
    error={false}
    fullWidth={true}
    inputProps={
      Object {
        "data-testid": "firstName-field",
        "ref": [Function],
      }
    }
    label="auth:firstName"
    margin="normal"
    name="firstName"
    type="text"
    variant="outlined"
  />
  <WithStyles(ForwardRef(TextField))
    error={false}
    fullWidth={true}
    inputProps={
      Object {
        "data-testid": "lastName-field",
        "ref": [Function],
      }
    }
    label="auth:lastName"
    margin="normal"
    name="lastName"
    type="text"
    variant="outlined"
  />
  <WithStyles(ForwardRef(TextField))
    error={false}
    fullWidth={true}
    inputProps={
      Object {
        "data-testid": "password-field",
        "ref": [Function],
      }
    }
    label="auth:password"
    margin="normal"
    name="password"
    type="password"
    variant="outlined"
  />
  <WithStyles(ForwardRef(TextField))
    error={false}
    fullWidth={true}
    inputProps={
      Object {
        "data-testid": "repeatPassword-field",
        "ref": [Function],
      }
    }
    label="auth:repeatPassword"
    margin="normal"
    name="repeatPassword"
    type="password"
    variant="outlined"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    control={
      <WithStyles(ForwardRef(Checkbox))
        color="primary"
        inputProps={
          Object {
            "data-testid": "checkbox",
          }
        }
        inputRef={[Function]}
        name="checked"
      />
    }
    label="auth:acceptTermsAndCondition"
  />
  <div
    className="submitButton"
  >
    <WithStyles(ForwardRef(Button))
      className="registerButton"
      color="primary"
      data-test="submit-button"
      data-testid="submit-button"
      disabled={false}
      type="submit"
      variant="contained"
    >
      auth:registerButtonLabel
    </WithStyles(ForwardRef(Button))>
  </div>
</form>
`;
